### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] False positive when asserting to type with nested optional properties

### ISSUE THREAD

**Issue by ehoogeveen-medweb** [typescript-eslint/typescript-eslint#12270]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play#ts=5.9.2&fileType=.tsx&code=C4TwDgpgBMAiD2BjAqgO1JAJssmCGw0AvFAEoSLwBOmAPAM7BUCWqA5gDRQCuqA1qngB3VAD4oAMigBvKAkTcAthHQB%2BAFwyoAMWqKNURi3YBuHqmaZNR1mygBfByYCwAKAzQ4SHPkJQSXijo4BDYuATQUrLySirABrKIADZ49PQAcnjK1ky2Dk5uHjDyPhEAgojAzPCo-loxymqasniV1agGAOTc4YSd%2BfYurm6UqIxQ1GzyddKOqcXevRBDo%2BOYzABmGzNz9AsoSys1a5sbAOrMwAAWFVU1detbUPOBpYS37ZL7b8tuj%2BeXG5tGoAOgacRmz2BqE03SWnS4IKRk3kYKQsXQAEICsN3CFvksLtcPvcAiUliTalJXocRsdgFB-kSgXdUAAmB6nZ57Gm%2BCDMylDJmAylstEKRoMkgtaGwnp8hFQJEglFIcUY4DYwZuOljBnC4nQgDMnK2QtOAuN6slkNarLl8MRyKoUzV4KxTiAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
type tDocUntypedUpdate = Record<string, unknown> & { Document?: { Form?: string; unid: string } };
type tDocUpdate = tDocUntypedUpdate & { Document?: { className: string } };
type tDocUpdateAction = { Document?: { action?: 'update' } };

const orgDoc = {} as tDocUpdate;
const diff = {} as tDocUpdate;
const diffWithAction = diff as tDocUpdateAction & tDocUpdate;
diffWithAction.Document = { action: 'update', ...orgDoc.Document! };
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertions": ["error"],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No warning.

### Actual Result

It warns on `diff as tDocUpdateAction & tDocUpdate` and autofixes it to just `diff`, which then results in TS error `2353: Object literal may only specify known properties, and 'action' does not exist in type '{ Form?: string | undefined; unid: string; } & { className: string; }'.`

### Additional Info

Regressed in version `8.59.0`. I tried the canary build to pick up https://github.com/typescript-eslint/typescript-eslint/pull/12257 and https://github.com/typescript-eslint/typescript-eslint/pull/12269, but it seems to be a separate issue.

**Comment by bradzacher** [comment]
This is a weird bug due to nested assignability of optional properties.

```ts
type T1 =
  {
    Document?: {
      className: string;
    }
  };

type T2 =
  {
    Document?: {
      className: string;
      action?: 'update'
    }
  };

declare const t1: T1;
const t2 = t1 as T2; /// error



type T3 =
  {
    className: string;
  };

type T4 =
  {
    className: string;
    action?: 'update';
  };

declare const t3: T3;
const t4 = t3 as T4;
/// no error
```
[playground](https://typescript-eslint.io/play#ts=5.9.2&fileType=.tsx&code=C4TwDgpgBAKgjFAvAWAFBSgbzRjARAewGMBXAWwgDtgB%2BALix1wyIBsBDAZ04Dl2KGnYACcAlpQDmAbiYYAvkzkzUaUJFgAmJE2zpchUhWr1Ge5my69%2BEQSPHTZudkWCiClEwHISYACbtgCE9HBT0lNDRfCAthaCJ3IShgOAZ4ZXjKROAtRCSELk1lCJVUNWgYAGZtPV1zDm4%2BASghMUlleSLS8HKAFmqMWtwLButbVoczKGdXdy8ff0DPdqhwkqiYuITgJIrUivStpL7c4CqCmB7lIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)
